syntax = "proto3";

package primarybackup;

service PrimaryBackup {
    rpc Write(WriteRequest) returns (WriteResponse) {}
    
    // when the backup comes up, it initiates a sync 
    // operation and gets it state up to date. 
    rpc Sync(HeartBeat) returns (stream WriteRequest) {}

    // return heartbeat with state INIT if there are 
    // pending items in the log. 
    rpc CheckSync(HeartBeat) returns (HeartBeat) {}
}

message HeartBeat {
    enum State {
        // can be used when doing the catch up work after failure
        INIT = 0;

        // when in normal operation
        READY = 1;
    }
    State state = 1;
}

message WriteRequest {
    int32 blk_address = 1;

    bytes buffer = 2;
}

message WriteResponse {
    enum Status {
        FAIL = 0;

        PASS = 1;
    }

    Status status = 1;
}